### ISSUE TITLE
Support for Covariates in time-moe?

### ISSUE THREAD

**Issue by TroyTzou** [Time-MoE/Time-MoE#59]
Hello, I am currently using time-moe for time-series forecasting of wind and solar power generation. After fine-tuning, the solar power predictions achieve approximately 95% accuracy under stable conditions. However, in certain scenarios, the prediction accuracy drops to around 76%.
During both model training and prediction, I have followed the examples by using only the 'power' metric. I am wondering if it is possible to incorporate covariates such as weather-related data into the model. I hypothesize that adding such external variables could potentially enhance the model's predictive performance during weather variations.
Could you please clarify whether time-moe supports the integration of covariates, and if so, provide guidance on the appropriate implementation methods?
Thank you for your assistance!
This translation maintains a professional tone, clearly articulates your observations and questions, and directly addresses the technical aspects of covariates in the context of time-moe.

power variation trend diagram of solar PV：

![Image](https://github.com/user-attachments/assets/70b60e3e-df6b-43ac-88e1-f0be2ffdfb65)

Prediction results in steady state：

![Image](https://github.com/user-attachments/assets/90180369-f5db-4f40-8463-097879b24b88)

Prediction results under changing weather conditions：

![Image](https://github.com/user-attachments/assets/8a8ede1c-3eea-473a-9bbb-b3520be32d1c)

**Comment by TroyTzou** [comment]
Thank you for your reply.I'm looking forward to getting this feature out there.

**Comment by Free007boy** [comment]
> Thank you for your interest in our work!
> 
> Currently, Time-MoE does not support covariate features. I am actively working on developing this functionality, although progress has been delayed due to time constraints. As soon as the development is completed, I will release it to the latest branch promptly.
> 
> Please feel free to contact us, if you have any questions.

-------------------------------------------------------------------------
Excited and very much looking forward to the feature release !

**Comment by 777nine** [comment]
> 您好，我目前正在使用 time-moe 进行风能和太阳能发电的时间序列预测。微调后，太阳能预测在稳定条件下达到约 95% 的准确率。但是，在某些情况下，预测准确率会下降到 72% 左右。在模型训练和预测过程中，我仅使用了 'power' 指标来遵循这些示例。我想知道是否有可能将协变量（例如与天气相关的数据）合并到模型中。我假设添加此类外部变量可能会增强模型在天气变化期间的预测性能。您能否澄清一下 time-moe 是否支持协变量的整合，如果是，请提供有关适当实现方法的指导？感谢您的帮助！此翻译保持了专业的语气，清楚地阐明了您的观察和问题，并直接解决了 time-moe 上下文中协变量的技术方面。
> 
> 太阳能光伏功率变化趋势图：
> 
> ![Image](https://github.com/user-attachments/assets/70b60e3e-df6b-43ac-88e1-f0be2ffdfb65)
> 
> 稳定状态下的预测结果：
> 
> ![Image](https://github.com/user-attachments/assets/90180369-f5db-4f40-8463-097879b24b88)
> 
> 天气变化下的预测结果：
> 
> ![Image](https://github.com/user-attachments/assets/8a8ede1c-3eea-473a-9bbb-b3520be32d1c)

请问方便告知您微调用了多大的数据量吗？训练了几个epoch？感觉您的微调结果很好，预测精度挺高的。我用一个（92131，7）的数据集微调3个epoch后，微调的模型和没有微调的模型相比没有什么改进.

**Comment by Maple728** [comment]
@777nine  Could you share the command you used for fine-tuning? It's possible that the learning rate you set is too low. I will update some parameters in `main.py` related to the learning rate and learning rate scheduler to improve fine-tuning.

**Comment by 777nine** [comment]
Thank you for your reply. Below are the command and parameter settings I have fine-tuned. Could you please take a look and let me know how to improve them?
****python main.py -d 1.jsonl
main.py
import argparse
from time_moe.runner import TimeMoeRunner


if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument('--data_path', '-d', type=str, help='Path to training data. (Folder contains data files, or data file)')
    parser.add_argument('--model_path', '-m', type=str, default='/root/autodl-tmp/time-moe', help='Path to pretrained model. Default: Maple728/TimeMoE-50M')
    parser.add_argument('--output_path', '-o', type=str, default='logs/time_moe')
    parser.add_argument('--max_length', type=int, default=1024)
    parser.add_argument('--learning_rate', type=float, default=5e-5, help='learning rate')
    parser.add_argument('--min_learning_rate', type=float, default=0, help='minimum learning rate')

    parser.add_argument('--train_steps', type=int, default=None, help='number of training steps')
    parser.add_argument('--num_train_epochs', type=float, default=3.0, help='number of training epochs')
    parser.add_argument('--normalization_method', type=str, choices=['none', 'zero', 'max'], default='zero', help='normalization method for sequence')

    parser.add_argument('--seed', type=int, default=9899, help='random seed')
    parser.add_argument('--attn_implementation', type=str, choices=['auto', 'eager', 'flash_attention_2'], default='auto', help='attention implementation')
    
    parser.add_argument('--lr_scheduler_type', type=str, choices=['constant', 'linear', 'cosine', 'constant_with_warmup'], default='constant', help='learning rate scheduler type')
    parser.add_argument('--warmup_ratio', type=float, default=0.0, help='warmup ratio')
    parser.add_argument('--warmup_steps', type=int, default=0, help='warmup steps')
    parser.add_argument('--weight_decay', type=float, default=0.1, help='weight decay')
    
    parser.add_argument('--global_batch_size', type=int, default=16, help='global batch size')
    parser.add_argument('--micro_batch_size', type=int, default=8, help='micro batch size per device')
    
    parser.add_argument('--precision', choices=['fp32', 'fp16', 'bf16'], type=str, default='fp16', help='precision mode (default: fp32)')
    parser.add_argument('--gradient_checkpointing', action='store_true', help='enable gradient checkpointing')
    parser.add_argument('--deepspeed', type=str, default=None, help='DeepSpeed config file path')

    parser.add_argument('--save_steps', type=int, default=None, help='number of steps to save model')
    parser.add_argument('--save_strategy', choices=['steps', 'epoch', 'no'], type=str, default='no', help='save strategy')
    parser.add_argument('--save_total_limit', type=int, default=None, help='limit the number of checkpoints')
    parser.add_argument('--save_only_model', action='store_true', help='save only model')

    parser.add_argument('--logging_steps', type=int, default=1, help='number of steps to log')
    parser.add_argument('--evaluation_strategy', choices=['steps', 'epoch', 'no'], type=str, default='no', help='evaluation strategy')
    parser.add_argument('--eval_steps', type=int, default=None, help='number of evaluation steps')

    parser.add_argument('--adam_beta1', type=float, default=0.9, help='adam beta1')
    parser.add_argument('--adam_beta2', type=float, default=0.95, help='adam beta2')
    parser.add_argument('--adam_epsilon', type=float, default=1e-8, help='adam epsilon')
    parser.add_argument('--max_grad_norm', type=float, default=1.0, help='max gradient norm')
    parser.add_argument('--dataloader_num_workers', type=int, default=4, help='number of workers for dataloader')

    args = parser.parse_args()

    if args.normalization_method == 'none':
        args.normalization_method = None

    runner = TimeMoeRunner(
        model_path=args.model_path,
        output_path=args.output_path,
        seed=args.seed,
    )

    runner.train_model(
        max_length=args.max_length,
        data_path=args.data_path,
        normalization_method=args.normalization_method,
        attn_implementation=args.attn_implementation,

        micro_batch_size=args.micro_batch_size,
        global_batch_size=args.global_batch_size,
        
        train_steps=args.train_steps,
        num_train_epochs=args.num_train_epochs,
        
        precision=args.precision,
        evaluation_strategy=args.evaluation_strategy,
        eval_steps=args.eval_steps,
        save_strategy=args.save_strategy,
        save_steps=args.save_steps,
        learning_rate=args.learning_rate,
        min_learning_rate=args.min_learning_rate,
        adam_beta1=args.adam_beta1,
        adam_beta2=args.adam_beta2,
        adam_epsilon=args.adam_epsilon,
        lr_scheduler_type=args.lr_scheduler_type,
        warmup_ratio=args.warmup_ratio,
        warmup_steps=args.warmup_steps,
        weight_decay=args.weight_decay,
        gradient_checkpointing=args.gradient_checkpointing,
        deepspeed=args.deepspeed,
        logging_steps=args.logging_steps,
        max_grad_norm=args.max_grad_norm,
        dataloader_num_workers=args.dataloader_num_workers,
        save_only_model=args.save_only_model,
        save_total_limit=args.save_total_limit,
    )

**Comment by iganggang** [comment]
@TroyTzou 
Hi Troy, I’m working on a similar PV/wind forecasting setup.
May I ask which dataset you used for the solar PV power series? If it’s public, could you share the dataset name or link？

**Comment by TroyTzou** [comment]
@iganggang Hello, I am using an internal dataset from the company, so it is not convenient to make it public
